### ISSUE TITLE
Slow running query using tedious

### ISSUE THREAD

**Issue by haswent** [tediousjs/tedious#1112]
I'm using https://github.com/tediousjs/node-mssql/ 

When using the default tedious driver there is one particular query that takes 25seconds to run through node-mssql whereas if I run it directly on SSMS it takes 200ms

When switching the driver in node-mssql from tedious to msnodesqlv8 it seems to solve the issue

### Behavior:

Would expect the run time to be comparable 

Running the queries and monitoring on SQL server profiler:

*node/mssql (tedious driver)*

![image](https://user-images.githubusercontent.com/50839406/84884954-46525e00-b08a-11ea-9c7f-6876ce68c5f4.png)

*SSMS*

![image](https://user-images.githubusercontent.com/50839406/84884964-4bafa880-b08a-11ea-872c-8e6c050d2eaa.png)

Not sure why but the query consistently takes 25seconds through node/mssql 

### Configuration:

**Connection Configuration**
```
const config = {
	user: DB_USER,
	password: DB_PASS,
	server: DB_HOST,
	requestTimeout: 300000,
	pool: {
		min: 5
	}
}
```

How I connect to run the query:

```
const dbc = new sql.connect(config);

const ClientDb = 'Client.dbo';
const StockDb = 'Stock.dbo';

export {
	dbc,
	sql,
	ClientDb,
	StockDb
}

import { dbc, sql, StockDb } from '../../config/db'
```

Then the actual query:

```
return dbc.then(pool => {
        return pool.request()
            .input('clientId', sql.Int, clientId)
            .input('startDate', sql.DateTime, new Date(startDate))
            .input('endDate', sql.DateTime, new Date(endDate))
            .query(`SELECT v.Id as VehicleId,
                            n.Name as Make, 
                            m.Name as Model, 
                            v.Variant as Variant,
                            v.Registration as Registration,
                            v.Price As Price,
                            v.OfferPrice as OfferPrice,
                            v.OfferDisplay as OfferDisplay,
                            d.Name as DealershipName,
                            v.StockDate as StockDate,
                            v.PurchaseStandIn as SIV,
                            s.Name as PurchaseSource,
                            DATEDIFF(day, StockDate, GETDATE()) as DaysInStock,
                            v.SoldDate as SoldDate,
                            v.SoldPrice as SoldPrice,
                            s1.Name as SoldSource
                    FROM  ${StockDb}.UsedVehicles v 
                    JOIN  ${StockDb}.Models m on m.Id = v.ModelId
                    JOIN  ${StockDb}.Makes n ON n.Id = m.MakeId
                    JOIN  ${StockDb}.Dealerships d on d.Id = v.DealershipId
                    LEFT JOIN ${StockDb}.UsedPriceSources s on s.Id = v.PurchasePriceSourceId
                    LEFT JOIN ${StockDb}.UsedPriceSources s1 on s1.Id = v.SoldPriceSourceId
                    WHERE v.Status = 3
                    AND v.ClientId = @clientId
                    AND v.SoldDate between @startDate AND @endDate
                    ORDER BY SoldDate ASC`)
    }).then(result => {
        return result.recordset;
    }).catch(err => {
        console.error(err);
    });
```

### Software versions

  * NodeJS: 10
  * node-mssql: 6.2, also tried the newer 7.0 which uses Tedious 8.3 
  * SQL Server: 2017

**Comment by arthurschreiber** [comment]
@haswent 👋

Thank you for this bug report. I don't see anything obvious that would cause this query to run that much slower. How many rows are returned by this query?

Could you try a few more things to help narrow down the problem? 🙇

Can you also provide the column type information?
Can you try removing selected columns one by one to try and see whether a specific column is causing this slow down?
Can you try running the same query directly using `tedious`, without `node-mssql`?

Thank you again! ❤️

**Comment by haswent** [comment]
I seem to have narrowed down the problem selects, removing these two from the select statement reduces the run time from 25s to ~100ms:

- s.Name as PurchaseSource,
- s1.Name as SoldSource

I simply removed the two rows from the select, but kept the LEFT JOIN in place and it dramatically reduced the run time 

The datatype for Name column that both selects use is `nvarchar(150)`

**Comment by arthurschreiber** [comment]
@haswent Thanks for narrowing this down!

I do have two more questions. 😅 

What happens if you _only_ select those two columns, and none of the other columns? Is the performance still bad?
How many rows are returned by your query?

**Comment by haswent** [comment]
Selecting just those two columns, the performance is ~40ms,
Returning about 280 rows of data.

I then added `v.Id` back into the query and the results were still fast, but as soon as i added `n.Name as Make` (also nvarchar(150)) into it, it slowed down again

**Comment by arthurschreiber** [comment]
This is so weird! 🤔 

Could you provide the sql to create a table that matches the minimal columns needed to reproduce? Plus, if the data contained in those columns is not sensitive, could you provide them as well?

**Comment by arthurschreiber** [comment]
> Not sure if the issue is due to the fact that in most scenarios's the value for `s.Name `and `s1.Name `will be `null ` - so the library sometimes parses nvarchar as text and other times it's a null value

That should not be an issue. 🙇‍♂️

**Comment by arthurschreiber** [comment]
> I then added `v.Id` back into the query and the results were still fast, but as soon as i added `n.Name as Make` (also nvarchar(150)) into it, it slowed down again

That is interesting as well, because to me, it seems like it's not caused by the fact that these columns are `nvarchar`. It must be due to some weird interaction of different column types or something like that. But I really can't think of any obvious cause for this. 😞 

Getting an easily reproducible case for this will be super important to track this down.

**Comment by haswent** [comment]
I've attached a minified/obfuscated version of the database that has the relevant columns/tables and the same dataset I've been testing on

[stockdb_min.zip](https://github.com/tediousjs/tedious/files/4799341/stockdb_min.zip)

**Comment by arthurschreiber** [comment]
Thank you, this is perfect! I'll try and see if I can reproduce the issue on my side.

**Comment by haswent** [comment]
> Thank you, this is perfect! I'll try and see if I can reproduce the issue on my side.

Great! thanks for your help, also below is a representative query with the parameters:

```
SELECT v.Id as VehicleId,
        n.Name as Make, 
        m.Name as Model, 
        v.Variant as Variant,
        v.Registration as Registration,
        v.Price As Price,
        v.OfferPrice as OfferPrice,
        v.OfferDisplay as OfferDisplay,
        d.Name as DealershipName,
        v.StockDate as StockDate,
        v.PurchaseStandIn as SIV,
        s.Name as PurchaseSource,
        DATEDIFF(day, StockDate, GETDATE()) as DaysInStock,
        v.SoldDate as SoldDate,
        v.SoldPrice as SoldPrice,
        s1.Name as SoldSource
FROM  UsedVehicles v 
JOIN  Models m on m.Id = v.ModelId
JOIN  Makes n ON n.Id = m.MakeId
JOIN  Dealerships d on d.Id = v.DealershipId
LEFT JOIN UsedPriceSources s on s.Id = v.PurchasePriceSourceId
LEFT JOIN UsedPriceSources s1 on s1.Id = v.SoldPriceSourceId
WHERE v.Status = 3
AND v.ClientId = 12
AND v.SoldDate between '2018-01-01' AND '2020-06-18'
ORDER BY SoldDate ASC
```

**Comment by arthurschreiber** [comment]
I'm not able to reproduce the issue. 😞  The query finishes succesfully in relatively short time (less than a second) on my machine.

I want to rule out this isn't an issue in `node-mssql`. Can you try running the following and see how that behaves?

```
const { Connection, Request } = require('tedious');

const connection = new Connection({
  "server": "localhost",
  "authentication": {
    "type": "default",
    "options": {
      "userName": "...",
      "password": "..."
    }
  },
  "options": {
    "database": "StockDB",
    "encrypt": false
  }
});

connection.connect((err) => {
  if (err) {
    throw err;
  }

  const request = new Request(`
    SELECT v.Id as VehicleId,
      n.Name as Make, 
      m.Name as Model, 
      v.Variant as Variant,
      v.Registration as Registration,
      v.Price As Price,
      v.OfferPrice as OfferPrice,
      v.OfferDisplay as OfferDisplay,
      d.Name as DealershipName,
      v.StockDate as StockDate,
      v.PurchaseStandIn as SIV,
      s.Name as PurchaseSource,
      DATEDIFF(day, StockDate, GETDATE()) as DaysInStock,
      v.SoldDate as SoldDate,
      v.SoldPrice as SoldPrice,
      s1.Name as SoldSource
    FROM  UsedVehicles v 
    JOIN  Models m on m.Id = v.ModelId
    JOIN  Makes n ON n.Id = m.MakeId
    JOIN  Dealerships d on d.Id = v.DealershipId
    LEFT JOIN UsedPriceSources s on s.Id = v.PurchasePriceSourceId
    LEFT JOIN UsedPriceSources s1 on s1.Id = v.SoldPriceSourceId
    WHERE v.Status = 3
    AND v.ClientId = 12
    AND v.SoldDate between '2018-01-01' AND '2020-06-18'
    ORDER BY SoldDate ASC
  `, () => {
    console.log('done');
    connection.close();
  });

  request.on('row', (...args) => {
    // console.log(...args);
  });

  connection.execSql(request);
});
```

**Comment by haswent** [comment]
The above query does indeed run quicker when running with tedious directly

I've created two scripts (attached), one node-mssql and one tedious and added timers

The results i got:
```
tedious: 896.474ms
node-mssql: 25102.669ms
node-mssql (no input vars): 25243.435ms
```

[scripts.zip](https://github.com/tediousjs/tedious/files/4799862/scripts.zip)

**Comment by haswent** [comment]
I did raise an issue originally with node-mssql, might need to re-open that issue and it seems likely its not tedious but something in node-mssql

https://github.com/tediousjs/node-mssql/issues/1056

**Comment by arthurschreiber** [comment]
> The above query does indeed run quicker when running with tedious directly

Yeah. This must be some issue in `node-mssql` then. Sorry about the trouble! I hope @dhensby can help you get this figured out. Thank you for providing all the test cases to narrow this down. 🙇‍♂️

**Comment by pixelspark** [comment]
Leaving this here in the hope it saves someone a bit of time:

Not sure if it is the same issue, but I encountered something similar today: a query is very slow from typeorm (which uses node-mssql and tedious under the hood, using the other driver did not make a difference), but very quick when copy-pasted and run from SSMS (replacing the parameter values with their literal values first)

Turns out, the reason appears to be that text parameters (`SomeColumn = @0`) seem to be inserted in the query as `nvarchar` typed strings. In my case the `SomeColumn` is `varchar` and due to the mismatch, indexes are not used anymore. I fixed this for now by doing `(SomeColumn = CAST(@0 AS VARCHAR(4000)))`. 

I am not sure where the type of parameters is defined, it might be worthwhile to fix this somwhere.
